File: src/formats/smi.ts
// SAMI (.smi) adapter.
//
// Format gotchas:
//  - HTML-like, TAGS ARE CASE-INSENSITIVE (<Sync Start=..>, <SYNC START=..>).
//  - START TIME ONLY (milliseconds). Like LRC, a cue's END is derived from the
//    NEXT <SYNC> start. A blank cue (&nbsp; or empty <P>) means "subtitle off":
//    it is NOT emitted as a cue, only used as the end boundary of the previous one.
//  - Multi-language tracks via <P Class=KRCC|ENCC|...>. First milestone edits the
//    most-populated Class only; the <HEAD>/<STYLE> and Class list are kept in meta
//    so we can round-trip.
//  - <br> -> "\n"; HTML entities decoded (and re-encoded on output).

import { newCueId, emptyDocument, type AssSpan, type SubtitleDocument } from "../types/subtitle";
import { categorizeTag, decodeTags, SMI_REPRESENTABLE, type LossCategory } from "./assTags";
import { assColorToHex } from "../utils/color";

const DEFAULT_TAIL = 3; // seconds, when a final cue has no closing boundary
const SYNC_RE = /<sync\b[^>]*\bstart\s*=\s*["']?(\d+)["']?[^>]*>/gi;
const CLASS_RE = /<p\b[^>]*\bclass\s*=\s*["']?([a-z0-9_-]+)["']?/i;

interface RawSync {
  startMs: number;
  className: string | null;
  text: string; // decoded plain text ("" for a blank/off marker)
}

export function parseSmi(raw: string): SubtitleDocument {
  const doc = emptyDocument("smi");
  const src = raw.replace(/\r\n/g, "\n");

  // Preserve <HEAD> (title + <STYLE>) verbatim for round-trip.
  const headMatch = src.match(/<head\b[^>]*>([\s\S]*?)<\/head>/i);
  if (headMatch) doc.meta.smiHead = headMatch[1].trim();

  // Collect every <SYNC> with its raw inner content.
  const syncs: RawSync[] = [];
  let m: RegExpExecArray | null;
  const matches: { index: number; startMs: number }[] = [];
  SYNC_RE.lastIndex = 0;
  while ((m = SYNC_RE.exec(src)) !== null) {
    matches.push({ index: m.index + m[0].length, startMs: Number(m[1]) });
  }
  for (let i = 0; i < matches.length; i++) {
    const inner = src.slice(matches[i].index, i + 1 < matches.length ? findTagStart(src, matches[i + 1].index) : src.length);
    const classMatch = inner.match(CLASS_RE);
    syncs.push({
      startMs: matches[i].startMs,
      className: classMatch ? classMatch[1] : null,
      text: smiInnerToText(inner),
    });
  }

  // Pick the dominant language Class (most non-blank entries).
  const classCounts = new Map<string, number>();
  for (const s of syncs) {
    if (s.text) classCounts.set(s.className ?? "", (classCounts.get(s.className ?? "") ?? 0) + 1);
  }
  let mainClass: string | null = null;
  let best = -1;
  for (const [cls, count] of classCounts) {
    if (count > best) {
      best = count;
      mainClass = cls || null;
    }
  }
  if (mainClass) doc.meta.smiMainClass = mainClass;
  // Record other classes so the export can keep emitting their P tags if desired.
  const allClasses = [...new Set(syncs.map((s) => s.className).filter(Boolean))] as string[];
  if (allClasses.length) doc.meta.smiClasses = allClasses.join(",");

  // Build cues from the main-class stream; blank entries close the previous cue.
  const stream = syncs.filter((s) => (s.className ?? null) === mainClass || s.className === null);
  for (let i = 0; i < stream.length; i++) {
    const cur = stream[i];
    if (!cur.text) continue; // blank = off marker, handled as a boundary below
    const start = cur.startMs / 1000;
    const next = stream[i + 1];
    const end = next ? next.startMs / 1000 : start + DEFAULT_TAIL;
    doc.cues.push({ id: newCueId(), start, end, text: cur.text });
  }
  return doc;
}

export function serializeSmi(doc: SubtitleDocument): string {
  const cls = doc.meta.smiMainClass ?? "UNKNOWNCC";
  const head =
    doc.meta.smiHead ??
    `<TITLE>Glyphline</TITLE>\n<STYLE TYPE="text/css"><!--\nP { font-family: sans-serif; color: white; }\n.${cls} { Name: Track; lang: ko-KR; }\n--></STYLE>`;

  const body: string[] = [];
  const cues = [...doc.cues].sort((a, b) => a.start - b.start);
  cues.forEach((cue, i) => {
    const startMs = Math.round(cue.start * 1000);
    // ASS cues carry inline override tags in assSpans → convert the representable
    // subset to SMI HTML. Plain cues just get entity-encoded text + <br>.
    const inner = cue.assSpans?.length ? spansToSmiHtml(cue.assSpans).html : textToSmi(cue.text);
    body.push(`<SYNC Start=${startMs}><P Class=${cls}>${inner}`);
    // Insert an off marker at this cue's end unless the next cue starts there.
    const next = cues[i + 1];
    const endMs = Math.round(cue.end * 1000);
    if (!next || Math.round(next.start * 1000) > endMs) {
      body.push(`<SYNC Start=${endMs}><P Class=${cls}>&nbsp;`);
    }
  });

  return `<SAMI>\n<HEAD>\n${head}\n</HEAD>\n<BODY>\n${body.join("\n")}\n</BODY>\n</SAMI>\n`;
}

// Find where the tag that produced `innerStart` actually begins (back up to '<').
function findTagStart(src: string, innerStart: number): number {
  const lt = src.lastIndexOf("<", innerStart);
  return lt === -1 ? innerStart : lt;
}

function smiInnerToText(inner: string): string {
  let t = inner;
  // Stop at the next </SYNC> or trailing tags we don't want.
  t = t.replace(/<\/?(sync|body|sami)\b[^>]*>/gi, "");
  t = t.replace(/<p\b[^>]*>/gi, ""); // opening <P ...>
  t = t.replace(/<\/p>/gi, "");
  t = t.replace(/<br\s*\/?>/gi, "\n");
  t = t.replace(/<[^>]+>/g, ""); // strip any remaining tags
  t = decodeEntities(t);
  return t.trim();
}

function textToSmi(text: string): string {
  return encodeEntities(text).replace(/\n/g, "<br>");
}

// ─── ASS spans -> SMI HTML ─────────────────────────────────────────────────────

interface SmiFormatState {
  b: boolean;
  i: boolean;
  u: boolean;
  s: boolean;
  color: string | null;
  face: string | null;
  size: string | null;
  drawing: boolean;
}

function freshState(): SmiFormatState {
  return { b: false, i: false, u: false, s: false, color: null, face: null, size: null, drawing: false };
}

/**
 * Convert ASS spans to SMI HTML. Representable formatting (bold/italic/underline/
 * strikeout, primary colour, font name/size) becomes HTML tags; everything else
 * (position, animation, karaoke, clip, drawing, …) is dropped and recorded so the
 * caller can warn about the loss. Each text run is wrapped independently to keep
 * the markup well-formed despite ASS's toggle-style tags.
 */
export function spansToSmiHtml(spans: AssSpan[]): { html: string; dropped: Set<LossCategory> } {
  const state = freshState();
  const dropped = new Set<LossCategory>();
  let html = "";

  for (const span of spans) {
    if (span.tags) applyTags(span.tags, state, dropped);
    if (state.drawing) continue; // drawing coordinates are not text
    const txt = encodeEntities(span.text).replace(/\\N/g, "<br>").replace(/\\h/g, "&nbsp;");
    if (!txt) continue;
    html += wrapRun(txt, state);
  }
  return { html, dropped };
}

function applyTags(block: string, state: SmiFormatState, dropped: Set<LossCategory>) {
  for (const { name, arg, known } of decodeTags(block)) {
    switch (name) {
      case "b": state.b = arg !== "0" && arg !== ""; break;
      case "i": state.i = arg !== "0" && arg !== ""; break;
      case "u": state.u = arg !== "0" && arg !== ""; break;
      case "s": state.s = arg !== "0" && arg !== ""; break;
      case "c":
      case "1c": state.color = assColorToHex(arg); break;
      case "fn": state.face = arg || null; break;
      case "fs": state.size = arg || null; break;
      case "p":
        state.drawing = Number(arg) > 0;
        if (state.drawing) dropped.add("drawing"); // the vector shape is lost
        break;
      case "pbo": break; // baseline offset for drawing — ignore quietly
      case "r": Object.assign(state, freshState()); break; // reset formatting
      default:
        if (!known || !SMI_REPRESENTABLE.has(name)) dropped.add(categorizeTag(name));
    }
  }
}

function wrapRun(txt: string, s: SmiFormatState): string {
  const open: string[] = [];
  const close: string[] = [];
  const fontAttrs: string[] = [];
  if (s.color) fontAttrs.push(`color="${s.color}"`);
  if (s.face) fontAttrs.push(`face="${s.face}"`);
  if (s.size) fontAttrs.push(`size="${s.size}"`);
  if (fontAttrs.length) {
    open.push(`<font ${fontAttrs.join(" ")}>`);
    close.unshift("</font>");
  }
  for (const tag of ["b", "i", "u", "s"] as const) {
    if (s[tag]) {
      open.push(`<${tag}>`);
      close.unshift(`</${tag}>`);
    }
  }
  return open.join("") + txt + close.join("");
}

/** Categories of ASS information that would be lost exporting this doc to SMI. */
export function smiExportLoss(doc: SubtitleDocument): LossCategory[] {
  const all = new Set<LossCategory>();
  for (const cue of doc.cues) {
    if (!cue.assSpans?.length) continue;
    for (const c of spansToSmiHtml(cue.assSpans).dropped) all.add(c);
  }
  return [...all];
}

function decodeEntities(s: string): string {
  return s
    .replace(/&nbsp;/gi, " ")
    .replace(/&lt;/gi, "<")
    .replace(/&gt;/gi, ">")
    .replace(/&quot;/gi, '"')
    .replace(/&#39;/g, "'")
    .replace(/&amp;/gi, "&");
}

function encodeEntities(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

File: native/Sources/GlyphlineCore/Formats/SMI.swift
// SAMI (.smi) adapter (ported from ../../src/formats/smi.ts).
//
//  - HTML-like, tags are CASE-INSENSITIVE.
//  - START TIME ONLY (ms): a cue's END comes from the NEXT <SYNC>. A blank cue
//    (&nbsp;/empty <P>) means "subtitle off" — a boundary, not a cue.
//  - Multi-language via <P Class=KRCC|ENCC|…>; we edit the dominant class and keep
//    <HEAD>/<STYLE> + the class list in meta for round-trip.
//  - <br> ↔ "\n"; HTML entities decoded/encoded.
//  - Lossy export: only \b \i \u \s \c \fn \fs survive as HTML; everything else is
//    dropped and reported via smiExportLoss so the UI can warn.

import Foundation

private let SMI_DEFAULT_TAIL = 3.0 // seconds when a final cue has no boundary
private let syncRegex = try! NSRegularExpression(
    pattern: #"<sync\b[^>]*\bstart\s*=\s*["']?(\d+)["']?[^>]*>"#, options: [.caseInsensitive])
private let classRegex = try! NSRegularExpression(
    pattern: #"<p\b[^>]*\bclass\s*=\s*["']?([a-z0-9_-]+)["']?"#, options: [.caseInsensitive])
private let headRegex = try! NSRegularExpression(
    pattern: #"<head\b[^>]*>([\s\S]*?)</head>"#, options: [.caseInsensitive])

private struct RawSync {
    var startMs: Int
    var className: String?
    var text: String // decoded plain text ("" for a blank/off marker)
}

public func parseSmi(_ raw: String) -> SubtitleDocument {
    var doc = SubtitleDocument.empty(.smi)
    let src = raw.replacingOccurrences(of: "\r\n", with: "\n")
    let ns = src as NSString
    let full = NSRange(location: 0, length: ns.length)

    // Preserve <HEAD> verbatim.
    if let h = headRegex.firstMatch(in: src, range: full) {
        doc.meta["smiHead"] = ns.substring(with: h.range(at: 1)).trimmed()
    }

    // Every <SYNC …> with the offset where its inner content starts.
    let syncMatches = syncRegex.matches(in: src, range: full)
    var syncs: [RawSync] = []
    for (i, m) in syncMatches.enumerated() {
        let innerStart = m.range.location + m.range.length
        let innerEnd: Int
        if i + 1 < syncMatches.count {
            // Back up to the '<' that opens the next <SYNC>.
            let nextInner = syncMatches[i + 1].range.location
            innerEnd = max(innerStart, nextInner)
        } else {
            innerEnd = ns.length
        }
        let inner = ns.substring(with: NSRange(location: innerStart, length: innerEnd - innerStart))
        let innerNS = inner as NSString
        var className: String?
        if let c = classRegex.firstMatch(in: inner, range: NSRange(location: 0, length: innerNS.length)) {
            className = innerNS.substring(with: c.range(at: 1))
        }
        syncs.append(RawSync(
            startMs: Int(ns.substring(with: m.range(at: 1))) ?? 0,
            className: className,
            text: smiInnerToText(inner)
        ))
    }

    // Dominant language class (most non-blank entries).
    var classCounts: [String: Int] = [:]
    for s in syncs where !s.text.isEmpty {
        classCounts[s.className ?? "", default: 0] += 1
    }
    var mainClass: String?
    var best = -1
    for (cls, count) in classCounts where count > best {
        best = count
        mainClass = cls.isEmpty ? nil : cls
    }
    if let mainClass { doc.meta["smiMainClass"] = mainClass }

    var seen = Set<String>()
    let allClasses = syncs.compactMap(\.className).filter { seen.insert($0).inserted }
    if !allClasses.isEmpty { doc.meta["smiClasses"] = allClasses.joined(separator: ",") }

    // Cues from the main-class stream; blanks are boundaries, not cues.
    let stream = syncs.filter { $0.className == mainClass || $0.className == nil }
    for (i, cur) in stream.enumerated() {
        if cur.text.isEmpty { continue }
        let start = Double(cur.startMs) / 1000
        let end = i + 1 < stream.count ? Double(stream[i + 1].startMs) / 1000 : start + SMI_DEFAULT_TAIL
        doc.cues.append(Cue(id: newCueId(), start: start, end: end, text: cur.text))
    }
    return doc
}

public func serializeSmi(_ doc: SubtitleDocument) -> String {
    let cls = doc.meta["smiMainClass"] ?? "UNKNOWNCC"
    let head = doc.meta["smiHead"] ?? """
    <TITLE>Glyphline</TITLE>
    <STYLE TYPE="text/css"><!--
    P { font-family: sans-serif; color: white; }
    .\(cls) { Name: Track; lang: ko-KR; }
    --></STYLE>
    """

    var body: [String] = []
    let cues = doc.cues.sorted { $0.start < $1.start }
    for (i, cue) in cues.enumerated() {
        let startMs = Int((cue.start * 1000).rounded())
        let inner = (cue.assSpans?.isEmpty == false)
            ? spansToSmiHtml(cue.assSpans!).html
            : textToSmi(cue.text)
        body.append("<SYNC Start=\(startMs)><P Class=\(cls)>\(inner)")
        // Off marker at this cue's end unless the next cue starts there.
        let endMs = Int((cue.end * 1000).rounded())
        let nextStartMs = i + 1 < cues.count ? Int((cues[i + 1].start * 1000).rounded()) : nil
        if nextStartMs == nil || nextStartMs! > endMs {
            body.append("<SYNC Start=\(endMs)><P Class=\(cls)>&nbsp;")
        }
    }

    return "<SAMI>\n<HEAD>\n\(head)\n</HEAD>\n<BODY>\n\(body.joined(separator: "\n"))\n</BODY>\n</SAMI>\n"
}

private func replaceAll(_ pattern: String, _ input: String, _ template: String) -> String {
    guard let re = RegexCache.get(pattern, options: [.caseInsensitive]) else { return input }
    let ns = input as NSString
    return re.stringByReplacingMatches(
        in: input, range: NSRange(location: 0, length: ns.length), withTemplate: template)
}

private func smiInnerToText(_ inner: String) -> String {
    var t = inner
    t = replaceAll(#"</?(sync|body|sami)\b[^>]*>"#, t, "")
    t = replaceAll(#"<p\b[^>]*>"#, t, "")
    t = replaceAll(#"</p>"#, t, "")
    t = replaceAll(#"<br\s*/?>"#, t, "\n")
    t = replaceAll(#"<[^>]+>"#, t, "") // strip remaining tags
    return decodeEntities(t).trimmed()
}

private func textToSmi(_ text: String) -> String {
    encodeEntities(text).replacingOccurrences(of: "\n", with: "<br>")
}

// ─── ASS spans → SMI HTML (lossy) ─────────────────────────────────────────────

private struct SmiFormatState {
    var b = false, i = false, u = false, s = false
    var color: String?
    var face: String?
    var size: String?
    var drawing = false
}

/// Convert ASS spans to SMI HTML. Representable formatting becomes HTML; the rest
/// (position, animation, karaoke, clip, drawing…) is dropped and recorded.
public func spansToSmiHtml(_ spans: [AssSpan]) -> (html: String, dropped: Set<LossCategory>) {
    var state = SmiFormatState()
    var dropped = Set<LossCategory>()
    var html = ""

    for span in spans {
        if let tags = span.tags { applySmiTags(tags, &state, &dropped) }
        if state.drawing { continue } // drawing coordinates are not text
        let txt = encodeEntities(span.text)
            .replacingOccurrences(of: "\\N", with: "<br>")
            .replacingOccurrences(of: "\\h", with: "&nbsp;")
        if txt.isEmpty { continue }
        html += wrapSmiRun(txt, state)
    }
    return (html, dropped)
}

private func applySmiTags(_ block: String, _ state: inout SmiFormatState, _ dropped: inout Set<LossCategory>) {
    for tag in decodeTags(block) {
        let arg = tag.arg
        switch tag.name {
        case "b": state.b = arg != "0" && !arg.isEmpty
        case "i": state.i = arg != "0" && !arg.isEmpty
        case "u": state.u = arg != "0" && !arg.isEmpty
        case "s": state.s = arg != "0" && !arg.isEmpty
        case "c", "1c": state.color = assColorToHex(arg)
        case "fn": state.face = arg.isEmpty ? nil : arg
        case "fs": state.size = arg.isEmpty ? nil : arg
        case "p":
            state.drawing = (Int(arg) ?? 0) > 0
            if state.drawing { dropped.insert(.drawing) } // the vector shape is lost
        case "pbo":
            break // baseline offset for drawing — ignore quietly
        case "r":
            state = SmiFormatState() // reset formatting
        default:
            if !tag.known || !SMI_REPRESENTABLE.contains(tag.name) {
                dropped.insert(categorizeTag(tag.name))
            }
        }
    }
}

private func wrapSmiRun(_ txt: String, _ s: SmiFormatState) -> String {
    var open: [String] = []
    var close: [String] = []
    var fontAttrs: [String] = []
    if let c = s.color { fontAttrs.append("color=\"\(c)\"") }
    if let f = s.face { fontAttrs.append("face=\"\(f)\"") }
    if let z = s.size { fontAttrs.append("size=\"\(z)\"") }
    if !fontAttrs.isEmpty {
        open.append("<font \(fontAttrs.joined(separator: " "))>")
        close.insert("</font>", at: 0)
    }
    for (flag, tag) in [(s.b, "b"), (s.i, "i"), (s.u, "u"), (s.s, "s")] where flag {
        open.append("<\(tag)>")
        close.insert("</\(tag)>", at: 0)
    }
    return open.joined() + txt + close.joined()
}

/// Categories of ASS information that would be lost exporting this doc to SMI.
public func smiExportLoss(_ doc: SubtitleDocument) -> [LossCategory] {
    var all = Set<LossCategory>()
    for cue in doc.cues {
        guard let spans = cue.assSpans, !spans.isEmpty else { continue }
        all.formUnion(spansToSmiHtml(spans).dropped)
    }
    return Array(all)
}

private func decodeEntities(_ s: String) -> String {
    var t = s
    t = replaceAll(#"&nbsp;"#, t, " ")
    t = replaceAll(#"&lt;"#, t, "<")
    t = replaceAll(#"&gt;"#, t, ">")
    t = replaceAll(#"&quot;"#, t, "\"")
    t = t.replacingOccurrences(of: "&#39;", with: "'")
    t = replaceAll(#"&amp;"#, t, "&")
    return t
}

private func encodeEntities(_ s: String) -> String {
    s.replacingOccurrences(of: "&", with: "&amp;")
        .replacingOccurrences(of: "<", with: "&lt;")
        .replacingOccurrences(of: ">", with: "&gt;")
}
